File: app/src/main/java/dev/ragnarok/fenrir/fragment/CreatePollFragment.java
package dev.ragnarok.fenrir.fragment;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputEditText;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.ActivityUtils;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.listener.TextWatcherAdapter;
import dev.ragnarok.fenrir.model.Poll;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.CreatePollPresenter;
import dev.ragnarok.fenrir.mvp.view.ICreatePollView;

public class CreatePollFragment extends BaseMvpFragment<CreatePollPresenter, ICreatePollView> implements ICreatePollView {

    public static final String REQUEST_CREATE_POLL = "request_create_poll";
    private TextInputEditText mQuestion;
    private CheckBox mAnonymous;
    private CheckBox mMultiply;
    private ViewGroup mOptionsViewGroup;

    public static CreatePollFragment newInstance(Bundle args) {
        CreatePollFragment fragment = new CreatePollFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static Bundle buildArgs(int accountId, int ownerId) {
        Bundle args = new Bundle();
        args.putInt(Extra.ACCOUNT_ID, accountId);
        args.putInt(Extra.OWNER_ID, ownerId);
        return args;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_create_poll, container, false);

        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));

        mQuestion = root.findViewById(R.id.dialog_poll_create_question);
        mAnonymous = root.findViewById(R.id.dialog_poll_create_anonymous);
        mMultiply = root.findViewById(R.id.dialog_poll_create_multiply);
        mOptionsViewGroup = root.findViewById(R.id.dialog_poll_create_options);

        for (int i = 0; i < mOptionsViewGroup.getChildCount(); i++) {
            TextInputEditText editText = (TextInputEditText) mOptionsViewGroup.getChildAt(i);
            int position = i;

            editText.addTextChangedListener(new TextWatcherAdapter() {
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    callPresenter(p -> p.fireOptionEdited(position, s));
                }

                @Override
                public void afterTextChanged(Editable s) {
                    if (TextUtils.isEmpty(s) || position == mOptionsViewGroup.getChildCount() - 1) {
                        return;
                    }

                    TextInputEditText next = (TextInputEditText) mOptionsViewGroup.getChildAt(position + 1);
                    if (next.getVisibility() == View.GONE) {
                        next.setVisibility(View.VISIBLE);
                    }
                }
            });
        }

        mQuestion.addTextChangedListener(new TextWatcherAdapter() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                callPresenter(p -> p.fireQuestionEdited(s));
            }
        });

        mAnonymous.setOnCheckedChangeListener((ignored, isChecked) -> callPresenter(p -> p.fireAnonyamousChecked(isChecked)));
        mMultiply.setOnCheckedChangeListener((ignored, isChecked) -> callPresenter(p -> p.fireMultiplyChecked(isChecked)));
        return root;
    }

    @NonNull
    @Override
    public IPresenterFactory<CreatePollPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> new CreatePollPresenter(
                requireArguments().getInt(Extra.ACCOUNT_ID),
                requireArguments().getInt(Extra.OWNER_ID),
                saveInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        ActionBar actionBar = ActivityUtils.supportToolbarFor(this);

        if (nonNull(actionBar)) {
            actionBar.setTitle(R.string.new_poll);
            actionBar.setSubtitle(null);
        }

        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(true)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
    }

    @Override
    public void displayQuestion(String question) {
        if (nonNull(mQuestion)) {
            mQuestion.setText(question);
        }
    }

    @Override
    public void setAnonymous(boolean anomymous) {
        if (nonNull(mAnonymous)) {
            mAnonymous.setChecked(anomymous);
        }
    }

    @Override
    public void setMultiply(boolean multiply) {
        if (nonNull(mMultiply)) {
            mMultiply.setChecked(multiply);
        }
    }

    @Override
    public void displayOptions(String[] options) {
        if (nonNull(mOptionsViewGroup)) {
            for (int i = 0; i < mOptionsViewGroup.getChildCount(); i++) {
                TextInputEditText editText = (TextInputEditText) mOptionsViewGroup.getChildAt(i);
                editText.setVisibility(View.VISIBLE);
                editText.setText(options[i]);
            }

            for (int u = mOptionsViewGroup.getChildCount() - 2; u >= 0; u--) {
                if (u == 1) {
                    break;
                }

                if (TextUtils.isEmpty(options[u])) {
                    mOptionsViewGroup.getChildAt(u + 1).setVisibility(View.GONE);
                } else {
                    break;
                }
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.add_menu) {
            callPresenter(CreatePollPresenter::fireDoneClick);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.add_menu, menu);
    }

    @Override
    public void showQuestionError(@StringRes int message) {
        if (nonNull(mQuestion)) {
            mQuestion.setError(getString(message));
            mQuestion.requestFocus();
        }
    }

    @Override
    public void showOptionError(int index, @StringRes int message) {
        if (nonNull(mOptionsViewGroup)) {
            ((TextInputEditText) mOptionsViewGroup.getChildAt(index)).setError(getString(message));
            mOptionsViewGroup.getChildAt(index).requestFocus();
        }
    }

    @Override
    public void sendResultAndGoBack(@NonNull Poll poll) {
        Bundle intent = new Bundle();
        intent.putParcelable(Extra.POLL, poll);
        getParentFragmentManager().setFragmentResult(REQUEST_CREATE_POLL, intent);

        requireActivity().onBackPressed();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/CreatePollFragment.kt
package dev.ragnarok.fenrir.fragment

import android.os.Bundle
import android.text.Editable
import android.view.*
import android.widget.CompoundButton
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuProvider
import com.google.android.material.checkbox.MaterialCheckBox
import com.google.android.material.textfield.TextInputEditText
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.ActivityUtils.supportToolbarFor
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.listener.TextWatcherAdapter
import dev.ragnarok.fenrir.model.Poll
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory
import dev.ragnarok.fenrir.mvp.presenter.CreatePollPresenter
import dev.ragnarok.fenrir.mvp.view.ICreatePollView

class CreatePollFragment : BaseMvpFragment<CreatePollPresenter, ICreatePollView>(),
    ICreatePollView, MenuProvider {
    private var mQuestion: TextInputEditText? = null
    private var mAnonymous: MaterialCheckBox? = null
    private var mMultiply: MaterialCheckBox? = null
    private var mOptionsViewGroup: ViewGroup? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_create_poll, container, false)
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        mQuestion = root.findViewById(R.id.dialog_poll_create_question)
        mAnonymous = root.findViewById(R.id.dialog_poll_create_anonymous)
        mMultiply = root.findViewById(R.id.dialog_poll_create_multiply)
        mOptionsViewGroup = root.findViewById(R.id.dialog_poll_create_options)
        mOptionsViewGroup?.let {
            for (i in 0 until it.childCount) {
                if (it.getChildAt(i) !is TextInputEditText) {
                    continue
                }
                val editText = it.getChildAt(i) as TextInputEditText
                editText.addTextChangedListener(object : TextWatcherAdapter() {
                    override fun onTextChanged(
                        s: CharSequence?,
                        start: Int,
                        before: Int,
                        count: Int
                    ) {
                        presenter?.fireOptionEdited(
                            i,
                            s
                        )
                    }

                    override fun afterTextChanged(s: Editable?) {
                        if (s.isNullOrEmpty() || i == it.childCount - 1) {
                            return
                        }
                        if (it.getChildAt(i + 1) is TextInputEditText) {
                            val next = it.getChildAt(i + 1) as TextInputEditText
                            if (next.visibility == View.GONE) {
                                next.visibility = View.VISIBLE
                            }
                        }
                    }
                })
            }
        }
        mQuestion?.addTextChangedListener(object : TextWatcherAdapter() {
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                presenter?.fireQuestionEdited(
                    s
                )
            }
        })
        mAnonymous?.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            presenter?.fireAnonyamousChecked(
                isChecked
            )
        }
        mMultiply?.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            presenter?.fireMultiplyChecked(
                isChecked
            )
        }
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(this, viewLifecycleOwner)
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<CreatePollPresenter> {
        return object : IPresenterFactory<CreatePollPresenter> {
            override fun create(): CreatePollPresenter {
                return CreatePollPresenter(
                    requireArguments().getInt(Extra.ACCOUNT_ID),
                    requireArguments().getInt(Extra.OWNER_ID),
                    saveInstanceState
                )
            }
        }
    }

    override fun onResume() {
        super.onResume()
        val actionBar = supportToolbarFor(this)
        if (actionBar != null) {
            actionBar.setTitle(R.string.new_poll)
            actionBar.subtitle = null
        }
        ActivityFeatures.Builder()
            .begin()
            .setHideNavigationMenu(true)
            .setBarsColored(requireActivity(), true)
            .build()
            .apply(requireActivity())
    }

    override fun displayQuestion(question: String?) {
        mQuestion?.setText(question)
    }

    override fun setAnonymous(anomymous: Boolean) {
        mAnonymous?.isChecked = anomymous
    }

    override fun setMultiply(multiply: Boolean) {
        mMultiply?.isChecked = multiply
    }

    override fun displayOptions(options: Array<String?>?) {
        mOptionsViewGroup?.let {
            for (i in 0 until it.childCount) {
                if (it.getChildAt(i) !is TextInputEditText) {
                    continue
                }
                val editText = it.getChildAt(i) as TextInputEditText
                editText.visibility = View.VISIBLE
                editText.setText(options?.get(i))
            }
            for (u in it.childCount - 2 downTo 0) {
                if (u == 1) {
                    break
                }
                if (options?.get(u).isNullOrEmpty()) {
                    it.getChildAt(u + 1).visibility = View.GONE
                } else {
                    break
                }
            }
        }
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem.itemId == R.id.add_menu) {
            presenter?.fireDoneClick()
            return true
        }
        return false
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.add_menu, menu)
    }

    override fun showQuestionError(@StringRes message: Int) {
        mQuestion?.error = getString(message)
        mQuestion?.requestFocus()
    }

    override fun showOptionError(index: Int, @StringRes message: Int) {
        (mOptionsViewGroup?.getChildAt(index) as TextInputEditText?)?.error = getString(message)
        mOptionsViewGroup?.getChildAt(index)?.requestFocus()
    }

    override fun sendResultAndGoBack(poll: Poll) {
        val intent = Bundle()
        intent.putParcelable(Extra.POLL, poll)
        parentFragmentManager.setFragmentResult(REQUEST_CREATE_POLL, intent)
        requireActivity().onBackPressedDispatcher.onBackPressed()
    }

    companion object {
        const val REQUEST_CREATE_POLL = "request_create_poll"
        fun newInstance(args: Bundle?): CreatePollFragment {
            val fragment = CreatePollFragment()
            fragment.arguments = args
            return fragment
        }

        fun buildArgs(accountId: Int, ownerId: Int): Bundle {
            val args = Bundle()
            args.putInt(Extra.ACCOUNT_ID, accountId)
            args.putInt(Extra.OWNER_ID, ownerId)
            return args
        }
    }
}